maskit/ensembles.py

```diff
@@ -0,0 +1,179 @@
+        executed for each ensemble branch in addition to one training step
+        that is done before the branching.
+        """
+        # first one trainingstep
+        params, _cost, _gradient = optimizer.step_cost_and_grad(
+            *args, masked_circuit.parameters, masked_circuit=masked_circuit
+        )
+        masked_circuit.parameters = params
+
+        # then branching
+        branches = self._branch(masked_circuit=masked_circuit)
+        basic_active_gates = masked_circuit.active().unwrap()
+        if branches is None:
+            return EnsembleResult(
+                branch=masked_circuit,
+                branch_name="center",
+                cost=args[0](
+                    masked_circuit.parameters, masked_circuit=masked_circuit
+                ).unwrap(),
+                gradient=_gradient,
+                brutto_steps=1,
+                netto_steps=1,
+                brutto=basic_active_gates,
+                netto=basic_active_gates,
+                ensemble=False,
+            )
+        branch_costs = []
+        branch_gradients = []
+        for branch in branches.values():
+            for _ in range(ensemble_steps):
+                params, _cost, _gradient = optimizer.step_cost_and_grad(
+                    *args, branch.parameters, masked_circuit=branch
+                )
+                branch.parameters = params
+            branch_costs.append(args[0](branch.parameters, masked_circuit=branch))
+            branch_gradients.append(_gradient)
+        minimum_index = branch_costs.index(min(branch_costs))
+        branch_name = list(branches.keys())[minimum_index]
+        selected_branch = branches[branch_name]
+        # FIXME: as soon as real (in terms of masked) parameters are used,
+        #   no mask has to be applied
+        #   until then real gradients must be calculated as gradients also
+        #   contain values from dropped gates
+        return EnsembleResult(
+            branch=selected_branch,
+            branch_name=branch_name,
+            cost=branch_costs[minimum_index].unwrap(),
+            gradient=branch_gradients[minimum_index],
+            brutto_steps=1 + len(branches) * ensemble_steps,
+            netto_steps=1 + ensemble_steps,
+            brutto=(
+                basic_active_gates
+                + sum(
+                    [branch.active() * ensemble_steps for branch in branches.values()]
+                )
+            ).unwrap(),
+            netto=(
+                basic_active_gates + selected_branch.active() * ensemble_steps
+            ).unwrap(),
+            ensemble=True,
+        )
+
+
+class IntervalEnsemble(Ensemble):
+    __slots__ = ("_interval", "_counter")
+
+    def __init__(self, dropout: Optional[Dict], interval: int):
+        super().__init__(dropout)
+        if interval < 1:
+            raise ValueError(
+                f"Interval must be in a valid range (>= 1), current value {interval}"
```

Comment:
Matching error messages. See below as well.
```suggestion
                f"interval must be >= 1, got {interval!r}"
```